bitcoin/src/pow.rs

```diff
@@ -913,6 +926,116 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+/// Error from `get_block_header_by_height` callback.
+///
+/// Occurse when there is an error while get a new block header
+/// from an external source.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
+
+impl fmt::Display for GetBlockHeaderError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match self {
+            Self::Generic(error) => error,
+            Self::NotFound => "Block header not found",
+        };
+        write!(f, "{msg}")
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for GetBlockHeaderError {}
+
+/// Calculate the next Work required as the `GetNextWorkRequired` do
+/// in bitcoin core.
+///
+/// Return a error if the callback provide from the caller fails.
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensus_params: Params,
+    current_height: Height,
+    current_header: Header,
+    new_block_timestamp: u32,
+    get_block_header_by_height: F,
+) -> Result<u32, GetBlockHeaderError>
+where
+    F: Fn(u32) -> Result<Header, GetBlockHeaderError>,
```

Comment:
FnMut is less restrictive, and since we are able to call it mutably, we should... unfortunately, since it can possibly be called multiple times on testnet, we can't use FnOnce.

```suggestion
    mut get_block_header_by_height: F,
) -> Result<u32, GetBlockHeaderError>
where
    F: FnMut(u32) -> Result<Header, GetBlockHeaderError>,
```

Reply:
I do not see why we should not do this, applying